src/ethereum/frontier/vm/instructions/memory.py

```diff
@@ -0,0 +1,119 @@
+"""
+Ethereum Virtual Machine (EVM) Memory Instructions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Implementations of the EVM Memory instructions.
+"""
+
+from ethereum.base_types import U256
+
+from .. import Evm
+from ..gas import GAS_BASE, GAS_VERY_LOW, subtract_gas
+from ..memory import (
+    extend_memory_and_subtract_gas,
+    memory_read_bytes,
+    memory_write,
+)
+from ..stack import pop, push
+
+
+def mstore(evm: Evm) -> None:
+    """
+    Stores a word to memory.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than
+        `GAS_LOW` + gas needed to extend memeory.
+    """
+    start_position = pop(evm.stack)
+    value = pop(evm.stack).to_be_bytes32()
+    extend_memory_and_subtract_gas(evm, start_position, U256(32))
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
```

Comment:
I prefer to keep it as is. Memory extension happens before writing to memory, so If there is an OutofGasException I'd like to follow the trace sequentially.

Reply:
Well, AFAIU I don't think it's economical to continue with memory expansion, given that you don't have enough Gas to even start the operation. It's a general pattern we follow in the rest of the opcode instructions as well. We first try to subtract the gas required for kicking off the operation, and later go to the internals of the operation.

\cc @SamWilsn @lightclient

Reply:
This is a spec, not a client. Readability triumphs over efficiency. For me reading the code in the order it occurs adds to better readability.

Reply:
I never mentioned anything about efficiency. For me it doesn't make sense to even reach the memory expansion bit if you don't even have sufficient gas to power the opcode.

My philosophy here is that you can't run a car unless it has enough power to at least start.

Reply:
My real reason for my approach is this section of the YP. 
<img width="720" alt="Screenshot 2021-07-15 at 11 25 25 PM" src="https://user-images.githubusercontent.com/8171193/125834820-a95fb8e6-3b76-49db-bbce-253ecee14533.png">
for better readability, I decided to have the gas logic as consecutive statements so that its easy to see the actual cost function for memory operations and that is very closely related to how the YP defines it.  
Up until now all the opcodes implemented had `Cmemory = 0`. So it was easier to have the gas logic right at the start of the method.

> My philosophy here is that you can't run a car unless it has enough power to at least start.

I'm going to nitpick on some code that you wrote for the sake of this discussion(No malice intended just healthy debate):
https://github.com/ethereum/eth1.0-specs/blob/308b8ced06daeefca14d1b2c47ff0336e0d6cd5f/src/ethereum/frontier/vm/instructions/arithmetic.py#L305-L315
Going by your philosophy, why didn't you do?:
```python
evm.gas_left = subtract_gas(evm.gas_left, GAS_EXPONENTIATION)

if exponent != 0:
     exponent_bits = exponent.bit_length()
     exponent_bytes = (exponent_bits + 7) // 8
     evm.gas_left = subtract_gas(evm.gas_left, GAS_EXPONENTIATION * exponent_bytes)
```

Reply:
@voith I don't think you understood my philosophy correctly. As my philosophy states, in the code that you highlighted, I am predicting the gas and deducting it previous to the actual opcode internals (which is to do the exponentiation).

But in your case, you are first calculating the cost needed for memory expansion and also expanding the memory which is an internal of the opcode, which is something none of our opcode implementations currently do (AFAIR). Let me know if I'm still not clear on my point.

I kindly request @SamWilsn @lightclient to opine on this issue. It's unclear how to proceed regarding this.

Reply:
If I understand this debate correctly, the PR [as currently written](https://github.com/ethereum/eth1.0-specs/blob/ae4b619f3b5848cdbea0428beb01242283bd15b5/src/ethereum/frontier/vm/memory.py) does the following:

 1. Checks if there is enough gas for memory expansion.
 2. Expands the memory.
 3. Checks if there is enough gas for the non-variable portion of `mstore`.

The disagreement here is specifically about (2) coming before (3)?

The arguments for switching (2) and (3)—so that we do all the gas accounting up front and expand the memory after—are:
 * Sane order of operations (it's a waste to start the car if it'll run out of fuel as soon as you try to move); and
 * Consistency with how other opcodes are implemented.

That seems to make sense to me, but I might be misinterpreting @voith's position.

In the grand scheme of things, I don't see this making a huge difference either way, and even if it does, I think it's something we can fix in the future.

This discussion is blocking quite a bit of code, so whatever gets us unblocked the quickest is my personal preference (and _please_ don't consider me as an ultimate arbiter for resolving PRs; I don't want that responsibility!)

Reply:
@SamWilsn @Bhargavasomu Take a look at d5b38d6d647cc6302a7e00889f9c6a6c4bb3cdca
I split the gas calculation method into a separate method. It's a bit more verbose but it exactly follows the yellow paper. 
Let me know what you think.